bleak/backends/bluezdbus/agent.py

```diff
@@ -0,0 +1,382 @@
+
+logger = logging.getLogger(__name__)
+
+# https://python-dbus-next.readthedocs.io/en/latest/type-system/index.html
+DBusObject = "o"
+DBusString = "s"
+DBusUInt16 = "q"
+DBusUInt32 = "u"
+
+
+class IOCapability(enum.Enum):
+    """I/O capabilities of this device, used for determining pairing method"""
+
+    DISPLAY_ONLY = "DisplayOnly"
+    DISPLAY_YES_NO = "DisplayYesNo"
+    KEYBOARD_ONLY = "KeyboardOnly"
+    NO_IO = "NoInputNoOutput"
+    KEYBOARD_DISPLAY = "KeyboardDisplay"
+
+
+class PairingAgentBlueZDBus(ServiceInterface):
+    """Agent for BlueZ pairing requests
+
+    Implemented by using the `BlueZ DBUS Agent API <https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc/agent-api.txt>`_.
+
+    Args:
+        io_capabilities (`Capability`): I/O capabilities of this device, used for determining pairing method.
+    """
+
+    def __init__(self, io_capabilities: IOCapability = IOCapability.KEYBOARD_DISPLAY):
+        super().__init__(defs.AGENT_INTERFACE)
+
+        # D-Bus message bus
+        self._bus: Optional[MessageBus] = None
+        # Path can be anything as long as it is unique
+        self._path = f"/org/bluez/agent{time.time() * 1000:.0f}"
+        # IO capabilities are required when the agent is registered
+        self._io_capabilities = io_capabilities
+
+        # Callback for every device (single agent handles all pairing requests)
+        self._callbacks: Dict[DBusObject, PairingCallback] = {}
+
+    def __del__(self) -> None:
+        if self._bus:
+            asyncio.ensure_future(self.unregister())
+
+    async def __aenter__(self) -> "PairingAgentBlueZDBus":
+        return await self.register()
+
+    async def __aexit__(self, *exc_info) -> None:
+        await self.unregister()
+
+    async def register(self) -> "PairingAgentBlueZDBus":
+        """
+        Register the Agent for handling pairing requests
+
+        Every application can register its own (and only one) agent and for all
+        actions triggered by that application its agent is used. If an application
+        chooses to not register an agent, the default system agent is used.
+        """
+        if self._bus:
+            # An application can only register one agent. Multiple agents per application is not
+            # supported and would raise error if not handled here.
+            return self
+        # Create system bus
+        bus = await MessageBus(bus_type=BusType.SYSTEM).connect()
+        # Make this Agent interface available on the given object path
+        bus.export(self._path, self)
+        # Register the Agent
+        reply = await bus.call(
+            Message(
+                destination=defs.BLUEZ_SERVICE,
+                path="/org/bluez",
+                interface=defs.AGENT_MANAGER_INTERFACE,
+                member="RegisterAgent",
+                signature="os",
+                body=[self._path, self._io_capabilities.value],
+            )
+        )
+        assert_reply(reply)
```

Comment:
I am now practically rewriting this feature while taking all these comments into account, and making it even more `bleak`-ish, e.g. using `class BleakPairingAgent(abc.ABC):` instead of callback.
However, while trying with example, I reached the following problem: agent must be registered before the connection is initiated, as example
```python
async with BleakPairingAgent() as pairing_agent, BleakClient(device) as client:
    ...
```
or
```python
async with BleakPairingAgent() as pairing_agent:
    async with BleakClient(device) as client:
        ...
```
in any case, I cannot use `_bus` from the client in the pairing agent... But I can do the other way around.

Reply:
Please base your work on the [bluez-adv-watcher](https://github.com/hbldh/bleak/tree/bluez-adv-watcher) branch from #884. It contains a new global BlueZManager that can be used to register the pairing agent independent of any BleakClient objects.